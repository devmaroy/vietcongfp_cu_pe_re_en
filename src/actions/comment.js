// Comment Actions

import database from '../firebase/firebase';
import history from '../routers/history';
import { 
    GET_COMMENTS,
    ADD_COMMENT,
    REMOVE_COMMENT,
    SET_COMMENT_LOADING,
    CLEAR_COMMENT_LOADING,
    GET_ERRORS
} from './types';


// Get Comments By Post Id
export const startGetCommentsByPostId = ( postId ) => {
    return ( dispatch ) => {
        dispatch( setCommentLoading() );

        return database
            .ref( 'comments' )
            .orderByChild( 'createdAt' )
            .once( 'value' )
            .then( ( snapshot ) => {
                const comments = [];
            
                snapshot.forEach( ( childSnapshot ) => {
                    if ( childSnapshot.val().postId === postId ) {
                        comments.push({
                            id: childSnapshot.key,
                            ...childSnapshot.val()
                        });
                    }
                });
    
                dispatch({
                    type: GET_COMMENTS,
                    comments
                });
    
                dispatch( clearCommentLoading() );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Add Comment
export const startAddComment = ( commentData ) => {
    return ( dispatch ) => {
        return database
            .ref( 'comments' ) 
            .push( commentData )
            .then( ( ref ) => {
                dispatch({
                    type: ADD_COMMENT,
                    comment: {
                        id: ref.key,
                        ...commentData
                    }
                });
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Remove Comment
export const startRemoveComment = ( id ) => {
    return ( dispatch ) => {
        return database 
            .ref( `comments/${ id }` )
            .remove()
            .then(() => {
                dispatch({
                    type: REMOVE_COMMENT,
                    id
                });
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Comment Loading
export const setCommentLoading = () => ({
    type: SET_COMMENT_LOADING
});


// Clear Comment Loading
export const clearCommentLoading = () => ({
    type: CLEAR_COMMENT_LOADING
});