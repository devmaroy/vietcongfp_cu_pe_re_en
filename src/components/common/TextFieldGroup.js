import React from 'react';
import PropTypes from 'prop-types';

const TextFieldGroup = ( { name, placeholder, value, error, info, type, onChange, disabled, className } ) => {
    return (
        <div>
            <input
                type={ type }
                placeholder={ placeholder }
                name={ name }
                value={ value }
                onChange={ onChange }
                disabled={ disabled }
                className={ className }
            />
            { info && <small>{ info }</small> }
            { error && <div>{ error }</div> }
        </div>
    );
};

TextFieldGroup.propTypes = {
    name: PropTypes.string.isRequired,
    placeholder: PropTypes.string,
    value: PropTypes.string.isRequired,
    info: PropTypes.string,
    error: PropTypes.string,
    type: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    disabled: PropTypes.bool,
    className: PropTypes.string
};
  
TextFieldGroup.defaultProps = {
    type: 'text'
};

export default TextFieldGroup;