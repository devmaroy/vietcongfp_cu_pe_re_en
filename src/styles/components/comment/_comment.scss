/*------------------------------------*\
    # Comment
\*------------------------------------*/


.comment-add {
    margin: $xl-size 0;
}

.comment {
    padding: $m-size 0;

    &:not( :last-child ) {
        border-bottom: .1rem dashed lighten( $light-green, 45% );
    }
}

.comment__title {
    color: $off-green;
}

.comment .star-icon {
    position: relative;
    top: .2rem;
    margin-left: .5rem;
}