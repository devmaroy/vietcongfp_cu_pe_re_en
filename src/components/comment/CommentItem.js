import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import moment from 'moment';
import { Link } from 'react-router-dom';
import { startRemoveComment } from '../../actions/comment';
import { createParagraphs } from '../../utils/text-helper-functions';

export const CommentItem = ( props ) => {
    const { comment, isAuthenticated, post } = props;
    const commentBody = createParagraphs( comment.body );
    const authorIcon = (
        <img 
            src="/images/icons/star.svg" 
            alt="Star icon"
            title="Author"
            className="icon star-icon"
        />
    )

    const onRemove = () => {
        props.startRemoveComment( comment.id );
    };

    return (
        <div className="comment">
            <h3
                className="comment__title"
            >
                { comment.name }
                { comment.name === post.name && authorIcon }
            </h3>

            <div className="comment__body">
                { commentBody }
            </div>

            <div className="comment_meta">
                <h5>{ moment( comment.createdAt ).format('MMMM Do YYYY, h:mm:ss a') }</h5>

                {
                    isAuthenticated && (
                        <button onClick={ onRemove } className="button">remove</button>
                    )
                }
            </div>
        </div>
    );
}

CommentItem.propTypes = {
    comment: PropTypes.object.isRequired,
    post: PropTypes.object,
    isAuthenticated: PropTypes.bool,
    startRemoveComment: PropTypes.func.isRequired
};

const mapStateToProps = ( state ) => ({
    isAuthenticated: state.auth.isAuthenticated,
    post: state.post.post
});

const mapDispatchToProps = ( dispatch ) => ({
    startRemoveComment: ( id ) => dispatch( startRemoveComment( id ) )
});

export default connect( mapStateToProps, mapDispatchToProps )( CommentItem );