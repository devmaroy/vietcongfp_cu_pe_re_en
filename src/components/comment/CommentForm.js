import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import TextFieldGroup from '../common/TextFieldGroup';
import TextAreaFieldGroup from '../common/TextAreaFieldGroup';
import Spinner from '../common/Spinner';
import history from '../../routers/history';

class CommentForm extends React.Component {
    state = {
        name: this.props.user ? this.props.user.name : '',
        body: '',
        errors: {}
    };
    
    handleValidation = () => {
        const errors = {};
        let formIsValid = true;

        // Name
        if ( ! this.state.name ) {
            formIsValid = false;
            errors[ 'name' ] = 'Name cannot be empty';
        }

        // Body
        if ( ! this.state.body ) {
            formIsValid = false;
            errors[ 'body' ] = 'Text cannot be empty';
        } else if ( this.state.body.length < 5 ) {
            formIsValid = false;
            errors[ 'body' ] = 'Text must be at least 5 characters.'
        }

        this.setState( () => ( { errors } ) );

        return formIsValid;
    };

    handleChange = ( e ) => {
        const target = e.target;
        const value = target.value;
        const name = target.name;
    
        this.setState( () => ( { [ name ]: value } ) );
    };

    onSubmit = ( e ) => {
        e.preventDefault();

        if ( this.handleValidation() ) {
            const commentData = {
                name: this.state.name,
                body: this.state.body,
                createdAt: Date.now(),
                postId: this.props.match.params.id
            };
            
            this.props.onSubmit( commentData );
            
            this.setState( () => ( { body: '' } ) );
        }
    };

    render() {
        const { errors } = this.state;
  
        return (
            <form onSubmit={ this.onSubmit }>
                <TextFieldGroup 
                    name="name"
                    placeholder="Name"
                    value={ this.state.name }
                    onChange={ this.handleChange }
                    error={ errors.name }
                    disabled={ this.props.isAuthenticated }
                />
                <TextAreaFieldGroup 
                    name="body"
                    placeholder="Body"
                    value={ this.state.body }
                    onChange={ this.handleChange }
                    error={ errors.body }
                />
                <button>Save Comment</button>
            </form>
        );
    };
};

CommentForm.propTypes = {
    user: PropTypes.object,
    isAuthenticated: PropTypes.bool
};

const mapStateToProps = ( state ) => ({
    user: state.auth.user,
    isAuthenticated: state.auth.isAuthenticated
});

export default connect( mapStateToProps )( withRouter( CommentForm ) );