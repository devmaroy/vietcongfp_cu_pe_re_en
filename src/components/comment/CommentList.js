import React from 'react';
import PropTypes from 'prop-types';
import Pagination from "react-js-pagination";
import CommentItem from './CommentItem';

class CommentList extends React.Component {
    state = {
        activePage: 1,
        commentsPerPage: 5
    }

    handlePageChange = ( pageNumber ) => {
        this.setState( () => ( { activePage: pageNumber } ) );
    }

    render() {
        const { activePage, commentsPerPage } = this.state;
        const { comments } = this.props;

        return (
            <div>
                <h3>Comments: { comments.length }</h3>
                {
                    comments.length === 0 ? (
                        <p>There are no comments.</p>
                    ) : (
                        <div>
                            <div className="comments">
                                {
                                    comments
                                        .slice( ( activePage - 1 ) * commentsPerPage, ( commentsPerPage * activePage ) )
                                        .map( ( comment ) => <CommentItem key={ comment.id } comment={ comment } /> )
                                }
                            </div>  
                            {  
                                comments.length > commentsPerPage && (
                                    <div className="site-pagination">
                                        <Pagination
                                            activePage={ this.state.activePage }
                                            itemsCountPerPage={ commentsPerPage }
                                            totalItemsCount={ comments.length }
                                            pageRangeDisplayed={ 5 }
                                            onChange={ this.handlePageChange }
                                        />
                                    </div>
                                )
                            }
                        </div>
                    )
                }
            </div>
        );
    };
};

CommentList.propTypes = {
    comments: PropTypes.array.isRequired
};

export default CommentList;