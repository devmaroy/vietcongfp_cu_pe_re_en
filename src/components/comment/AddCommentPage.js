import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import CommentForm from './CommentForm';
import { startAddComment } from '../../actions/post';

class AddCommentPage extends React.Component {
    onSubmit = ( commentData ) => {
        this.props.startAddComment( commentData );
    };

    render() {
        return (
            <div>
                <CommentForm 
                    onSubmit={ this.onSubmit }
                />
            </div>
        );
    };
};

AddCommentPage.propTypes = {
    startAddComment: PropTypes.func.isRequired
};

const mapDispatchToProps = ( dispatch ) => ({
    startAddComment: ( commentData ) => dispatch( startAddComment( commentData ) )
});

export default connect( undefined, mapDispatchToProps )( AddCommentPage );