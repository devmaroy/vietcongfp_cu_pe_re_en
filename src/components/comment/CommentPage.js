import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { startGetCommentsByPostId } from '../../actions/post';
import CommentList from './CommentList';
import AddCommentPage from '../comment/AddCommentPage';
import Spinner from '../common/Spinner';

class CommentPage extends React.Component {
    componentDidMount() {
        this.props.startGetCommentsByPostId( this.props.postId );
    }

    render() {
        const { comments, isCommentLoading } = this.props.post;
        let commentsContent;
        
        if ( comments === null || isCommentLoading ) {
            commentsContent = <Spinner />;

        } else {
            commentsContent = <CommentList comments={ comments } />;
        }

        return (
            <div>
                <AddCommentPage />
                { commentsContent }
            </div>
        );
    };
};

CommentPage.propTypes = {
    post: PropTypes.object.isRequired,
    startGetCommentsByPostId: PropTypes.func.isRequired
};

const mapStateToProps = ( state ) => ({
    post: state.post
});

const mapDispatchToProps = ( dispatch ) => ({
    startGetCommentsByPostId: ( postId ) => dispatch( startGetCommentsByPostId( postId ) )
});

export default connect( mapStateToProps, mapDispatchToProps )( CommentPage );