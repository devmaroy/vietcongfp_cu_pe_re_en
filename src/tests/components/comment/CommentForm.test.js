import React from 'react';
import { shallow } from 'enzyme';
import { CommentForm } from '../../../components/comment/CommentForm';
import TextFieldGroup from '../../../components/common/TextFieldGroup';
import TextAreaFieldGroup from '../../../components/common/TextAreaFieldGroup';
import comments from '../../fixtures/comments';
import moment from 'moment';

test( 'should correctly render form inside PostForm', () => {
    const wrapper = shallow(
        <CommentForm
            onSubmit={ () => {} }
        />
    );

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find( 'form' ).exists() ).toBeTruthy();
});

test( 'should correctly render error object if no data provided', () => {
    const wrapper = shallow(
        <CommentForm
            onSubmit={ () => {} }
        />
    );

    wrapper.find( 'form' ).simulate( 'submit', {
        preventDefault: () => {}
    });

    expect( wrapper.state( 'errors' ).name.length ).toBeGreaterThan( 0 );
    expect( wrapper.state( 'errors' ).body.length ).toBeGreaterThan( 0 );
}); 

test( 'should set name on change', () => {
    const wrapper = shallow(
        <CommentForm
            onSubmit={ () => {} }
        />
    );

    const value = 'Phili480';
    const event = { target: { name: 'name', value } };

    wrapper.find( TextFieldGroup ).simulate( 'change', event );

    expect( wrapper.state( 'name' ) ).toBe( value );
});

test( 'should set body on change', () => {
    const wrapper = shallow(
        <CommentForm
            onSubmit={ () => {} }
        />
    );

    const value = 'some nice comment here lol';
    const event = { target: { name: 'body', value } };

    wrapper.find( TextAreaFieldGroup ).simulate( 'change', event );

    expect( wrapper.state( 'body' ) ).toBe( value );
});

test( 'should call onSubmit prop for valid form submission', () => {
    const onSubmitSpy = jest.fn(); 

    const wrapper = shallow(
        <CommentForm
            onSubmit={ onSubmitSpy }
            match={ { params: { id: "1" } } }
        />
    );

    wrapper.setState( { name: comments[0].name, body: comments[0].body } );

    wrapper.find( 'form' ).simulate( 'submit', {
        preventDefault: () => {}
    });

    expect( wrapper.state( 'errors' ) ).toEqual( {} )

    expect( onSubmitSpy ).toHaveBeenLastCalledWith({
        name: comments[0].name,
        body: comments[0].body,
        createdAt: moment().valueOf(),
        postId: '1'
    });
});