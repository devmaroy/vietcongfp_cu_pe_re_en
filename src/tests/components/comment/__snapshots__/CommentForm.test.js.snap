// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render form inside PostForm 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="input-group"
  >
    <div
      className="input-group__item"
    >
      <TextFieldGroup
        className="text-input"
        name="name"
        onChange={[Function]}
        placeholder="Name"
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="input-group"
  >
    <div
      className="input-group__item"
    >
      <TextAreaFieldGroup
        className="textarea"
        name="body"
        onChange={[Function]}
        placeholder="Your message"
        value=""
      />
    </div>
  </div>
  <div
    className="form-meta"
  >
    <button
      className="button"
    >
      Save Comment
    </button>
  </div>
</form>
`;
